File: packages/tool-presentation/src/beeui.ts
import { z } from "zod";

import {
  questionComponentSchema,
  renderBeeQuestion,
  type BeeQuestion,
} from "./bee-question";
import { scrubIdentifiers } from "./scrub-identifiers";

// The single source of truth for Bee's generative-UI ("beeui") contract.
// Every client — web, mobile, CLI, iMessage — parses fenced beeui blocks
// through this module so the vocabulary can never drift between channels.

const httpsUrlSchema = z
  .string()
  .url()
  .refine((url) => url.startsWith("https://"), "Expected an HTTPS URL");

export const firstFocusPreviewSchema = z.object({
  type: z.literal("first_focus"),
  requestId: z.string().min(1),
  goalTitle: z.string().min(1),
  projectTitle: z.string().min(1),
  taskTitle: z.string().min(1),
  seed: z.string().min(1).optional(),
  highlightExpiresAt: z.number().finite().optional(),
});

export type FirstFocusPreview = z.infer<typeof firstFocusPreviewSchema>;

export const uiComponentSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("text"), body: z.string() }),
  z.object({
    type: z.literal("metric"),
    label: z.string(),
    value: z.string(),
    delta: z.string().optional(),
  }),
  z.object({
    type: z.literal("chart"),
    kind: z.literal("bar"),
    title: z.string(),
    unit: z.string().optional(),
    data: z.array(z.object({ label: z.string(), value: z.number() })).min(1),
  }),
  z.object({
    type: z.literal("tasks"),
    title: z.string(),
    items: z.array(
      z.object({
        id: z.string(),
        title: z.string(),
        done: z.boolean(),
        due: z.string().optional(),
      }),
    ),
  }),
  z.object({
    type: z.literal("highlight"),
    title: z.string(),
    body: z.string(),
  }),
  z.object({
    type: z.literal("image"),
    url: httpsUrlSchema,
    alt: z.string(),
    title: z.string().optional(),
  }),
  z.object({
    type: z.literal("bookmark"),
    title: z.string(),
    url: httpsUrlSchema,
    kind: z.enum(["website", "tweet", "youtube"]).optional(),
    labels: z.array(z.string()).max(8).optional(),
    note: z.string().optional(),
  }),
  z.object({
    type: z.literal("devin"),
    title: z.string(),
    status: z.string(),
    statusDetail: z.string().optional(),
    sessionId: z.string().regex(/^devin-[A-Za-z0-9_-]+$/),
    sessionUrl: httpsUrlSchema,
    summary: z.string().optional(),
    pullRequests: z
      .array(z.object({ url: httpsUrlSchema, state: z.string().optional() }))
      .max(20),
  }),
  firstFocusPreviewSchema,
  z.object({
    type: z.literal("confirm"),
    summary: z.string(),
    action: z.string(),
    payload: z.record(z.string(), z.unknown()).optional(),
  }),
  questionComponentSchema,
]);

export type UIComponent = z.infer<typeof uiComponentSchema>;

/** A component type this client build does not know; degrade, never leak JSON. */
export type UnsupportedComponent = { type: "unsupported" };

export type ParsedBeeUiComponent = UIComponent | UnsupportedComponent;

const KNOWN_COMPONENT_TYPES = new Set<string>(
  uiComponentSchema.options.map((option) => option.shape.type.value),
);

/** Matches an opening beeui fence, e.g. to hide a block that is still streaming. */
export const BEEUI_FENCE_OPEN = /```beeui/i;

const beeUiFence = () => /```beeui\s*([\s\S]*?)```/gi;
const markdownImage = () =>
  /!\[([^\]]*)\]\((https:\/\/[^\s)]+)(?:\s+["'][^"']*["'])?\)/g;

/** Machine ids belong in structured fields, never in copy the user reads. */
function scrubComponent(component: UIComponent): UIComponent {
  switch (component.type) {
    case "text":
      return { ...component, body: scrubIdentifiers(component.body) };
    case "metric":
      return {
        ...component,
        label: scrubIdentifiers(component.label),
        value: scrubIdentifiers(component.value),
        delta: component.delta
          ? scrubIdentifiers(component.delta)
          : component.delta,
      };
    case "chart":
    case "tasks":
      return { ...component, title: scrubIdentifiers(component.title) };
    case "highlight":
      return {
        ...component,
        title: scrubIdentifiers(component.title),
        body: scrubIdentifiers(component.body),
      };
    case "image":
      return {
        ...component,
        alt: scrubIdentifiers(component.alt),
        title: component.title
          ? scrubIdentifiers(component.title)
          : component.title,
      };
    case "devin":
      return {
        ...component,
        title: scrubIdentifiers(component.title),
        summary: component.summary
          ? scrubIdentifiers(component.summary)
          : component.summary,
      };
    case "bookmark":
      return {
        ...component,
        title: scrubIdentifiers(component.title),
        note: component.note ? scrubIdentifiers(component.note) : component.note,
      };
    case "confirm":
      return { ...component, summary: scrubIdentifiers(component.summary) };
    case "question":
      return {
        ...component,
        questions: component.questions.map((question) => ({
          ...question,
          header: scrubIdentifiers(question.header),
          question: scrubIdentifiers(question.question),
          options: question.options?.map((option) => ({
            ...option,
            label: scrubIdentifiers(option.label),
            description: option.description
              ? scrubIdentifiers(option.description)
              : option.description,
          })),
        })),
      };
    default:
      return component;
  }
}

function record(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : undefined;
}

/**
 * Parses one fenced beeui JSON payload. Component types this build does not
 * know degrade to an "unsupported" card so newer agent vocabulary never
 * silently disappears; an invalid *known* component drops the entire block so
 * a malformed generated payload can never leak raw JSON into the chat.
 */
export function parseBeeUiBlock(json: string): ParsedBeeUiComponent[] {
  try {
    const payload = record(JSON.parse(json));
    if (!payload || !Array.isArray(payload.components)) return [];
    const components: ParsedBeeUiComponent[] = [];
    for (const value of payload.components) {
      const type = record(value)?.type;
      if (typeof type !== "string") return [];
      if (!KNOWN_COMPONENT_TYPES.has(type)) {
        components.push({ type: "unsupported" });
        continue;
      }
      const parsed = uiComponentSchema.safeParse(value);
      if (!parsed.success) return [];
      components.push(scrubComponent(parsed.data));
    }
    return components;
  } catch {
    return [];
  }
}

/**
 * Splits Bee's reply into conversational copy and validated UI components.
 * Specialists may return Markdown images directly; those are promoted into
 * image cards so every channel can make its own rendering decision.
 */
export function extractBeeUi(text: string): {
  spoken: string;
  components: ParsedBeeUiComponent[];
} {
  const components: ParsedBeeUiComponent[] = [];
  const spoken = text
    .replace(beeUiFence(), (_match, json: string) => {
      components.push(...parseBeeUiBlock(json.trim()));
      return "";
    })
    .replace(markdownImage(), (_match, alt: string, url: string) => {
      if (
        !components.some(
          (component) => component.type === "image" && component.url === url,
        )
      ) {
        components.push({
          type: "image",
          url,
          alt: alt.trim() || "Generated image",
        });
      }
      return "";
    })
    // Keep line breaks so markdown structure survives; just trim the excess.
    .replace(/[ \t]+\n/g, "\n")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
  return { spoken: scrubIdentifiers(spoken), components };
}

export type Web3Confirmation = {
  actionId: string;
  summary: string;
};

export type BeeUiFollowUps = {
  firstFocus?: FirstFocusPreview;
  web3Confirmation?: Web3Confirmation;
  confirmation?: { summary: string };
  question?: BeeQuestion;
};

function web3ConfirmationOf(
  component: ParsedBeeUiComponent,
): Web3Confirmation | undefined {
  if (component.type !== "confirm") return undefined;
  const actionId = component.payload?.web3ActionId;
  return typeof actionId === "string" && actionId
    ? { actionId, summary: component.summary }
    : undefined;
}

/**
 * Derives the actionable follow-ups a text channel can answer in-thread.
 * A Web3 confirmation is only surfaced when the reply contains exactly one
 * action-bound confirm card — never guess which funds movement was meant.
 */
export function deriveBeeUiFollowUps(
  components: ParsedBeeUiComponent[],
): BeeUiFollowUps {
  const firstFocus = components.find(
    (component): component is FirstFocusPreview =>
      component.type === "first_focus",
  );
  const web3Confirmations = components
    .map(web3ConfirmationOf)
    .filter((confirmation): confirmation is Web3Confirmation =>
      Boolean(confirmation),
    );
  // The latest blocking decision wins: agent envelopes accumulate steps, so
  // earlier drafts must not compete with the final card.
  const reversed = [...components].reverse();
  const confirmation = reversed.find(
    (component): component is Extract<UIComponent, { type: "confirm" }> =>
      component.type === "confirm" && !web3ConfirmationOf(component),
  );
  const question = reversed.find(
    (component): component is Extract<UIComponent, { type: "question" }> =>
      component.type === "question",
  );
  return {
    ...(firstFocus ? { firstFocus } : {}),
    ...(web3Confirmations.length === 1
      ? { web3Confirmation: web3Confirmations[0] }
      : {}),
    ...(confirmation ? { confirmation: { summary: confirmation.summary } } : {}),
    ...(question ? { question: { questions: question.questions } } : {}),
  };
}

export type BeeUiMarkdown = {
  markdown: string;
  links: string[];
};

/**
 * Renders one component as channel-neutral Markdown for text channels
 * (CLI, iMessage). Rich clients render components natively instead.
 */
export function renderBeeUiMarkdown(
  component: ParsedBeeUiComponent,
): BeeUiMarkdown {
  switch (component.type) {
    case "text":
      return { markdown: component.body, links: [] };
    case "metric":
      return {
        markdown: `**${component.label}:** ${component.value}${
          component.delta ? ` — ${component.delta}` : ""
        }`,
        links: [],
      };
    case "chart":
      return {
        markdown: [
          `**${component.title}**`,
          ...component.data.map(
            (item) =>
              `${item.label}: ${item.value}${
                component.unit ? ` ${component.unit}` : ""
              }`,
          ),
        ].join("\n"),
        links: [],
      };
    case "tasks":
      return {
        markdown: [
          `**${component.title}**`,
          ...component.items.map(
            (item) =>
              `${item.done ? "☑" : "☐"} ${item.title}${
                item.due ? ` — ${item.due}` : ""
              }`,
          ),
          "Reply with the exact Task you want Bee to work with.",
        ].join("\n"),
        links: [],
      };
    case "highlight":
      return {
        markdown: `**${component.title}**\n${component.body}`,
        links: [],
      };
    case "image":
      return {
        markdown: component.title ? `**${component.title}**` : component.alt,
        links: [component.url],
      };
    case "bookmark":
      return {
        markdown: `**${component.title}**${
          component.note ? `\n${component.note}` : ""
        }`,
        links: [component.url],
      };
    case "devin":
      return {
        markdown: [
          `**${component.title}** — ${component.status}`,
          component.statusDetail ?? "",
          component.summary ?? "",
          ...component.pullRequests.map(
            (pullRequest) =>
              `Pull request${pullRequest.state ? ` — ${pullRequest.state}` : ""}`,
          ),
        ]
          .filter(Boolean)
          .join("\n"),
        links: [
          component.sessionUrl,
          ...component.pullRequests.map((pullRequest) => pullRequest.url),
        ],
      };
    case "first_focus":
      return {
        markdown: [
          "**Your first focus**",
          `Goal: ${component.goalTitle}`,
          `Project: ${component.projectTitle}`,
          `Task: ${component.taskTitle}`,
          "Reply **yes** to create it or **no** to cancel.",
        ].join("\n"),
        links: [],
      };
    case "confirm":
      return {
        markdown: [
          "**Needs your confirmation**",
          component.summary,
          web3ConfirmationOf(component)
            ? "Reply **yes** to authorize this exact action or **no** to cancel it."
            : "Reply **yes** to continue or **no** to cancel.",
        ].join("\n"),
        links: [],
      };
    case "question":
      return {
        markdown: renderBeeQuestion({ questions: component.questions }),
        links: [],
      };
    case "unsupported":
      return {
        markdown:
          "Bee shared an interactive card that can’t be displayed here. Open BeeGreat to continue.",
        links: ["https://beegreat.app"],
      };
  }
}

File: apps/android/core/contract/src/main/kotlin/com/beegreat/contract/BeeUi.kt
package com.beegreat.contract

import java.net.URI
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull

/**
 * Bee's generative-UI vocabulary. Port of `packages/tool-presentation/src/beeui.ts`,
 * the single source for every client (ADR 0003). Rules:
 * - an unknown component type degrades to [BeeUiComponent.Unsupported];
 * - an invalid known component drops the entire block;
 * - machine ids are scrubbed from every user-visible string.
 */
sealed interface BeeUiComponent {
  data class Text(val body: String) : BeeUiComponent

  data class Metric(val label: String, val value: String, val delta: String? = null) : BeeUiComponent

  data class Chart(val title: String, val unit: String? = null, val data: List<Bar>) : BeeUiComponent {
    data class Bar(val label: String, val value: Double)
  }

  data class Tasks(val title: String, val items: List<Item>) : BeeUiComponent {
    data class Item(val id: String, val title: String, val done: Boolean, val due: String? = null)
  }

  data class Highlight(val title: String, val body: String) : BeeUiComponent

  data class Image(val url: String, val alt: String, val title: String? = null) : BeeUiComponent

  data class Bookmark(
    val title: String,
    val url: String,
    val kind: String? = null,
    val labels: List<String>? = null,
    val note: String? = null,
  ) : BeeUiComponent

  data class Devin(
    val title: String,
    val status: String,
    val statusDetail: String? = null,
    val sessionId: String,
    val sessionUrl: String,
    val summary: String? = null,
    val pullRequests: List<PullRequest>,
  ) : BeeUiComponent {
    data class PullRequest(val url: String, val state: String? = null)
  }

  data class FirstFocus(
    val requestId: String,
    val goalTitle: String,
    val projectTitle: String,
    val taskTitle: String,
    val seed: String? = null,
    val highlightExpiresAt: Double? = null,
  ) : BeeUiComponent

  data class Confirm(val summary: String, val action: String, val payload: JsonObject? = null) : BeeUiComponent

  data class Question(val questions: List<Prompt>) : BeeUiComponent {
    data class Prompt(val header: String, val question: String, val options: List<Option>? = null)

    data class Option(val label: String, val description: String? = null)
  }

  data object Unsupported : BeeUiComponent
}

data class BeeUiExtraction(val spoken: String, val components: List<BeeUiComponent>)

data class Web3Confirmation(val actionId: String, val summary: String)

data class BeeUiFollowUps(
  val firstFocus: BeeUiComponent.FirstFocus? = null,
  val web3Confirmation: Web3Confirmation? = null,
  val confirmationSummary: String? = null,
  val question: BeeUiComponent.Question? = null,
)

/** Matches an opening beeui fence, e.g. to hide a block that is still streaming. */
val BEEUI_FENCE_OPEN = Regex("```beeui", RegexOption.IGNORE_CASE)

private val BEEUI_FENCE = Regex("```beeui\\s*([\\s\\S]*?)```", RegexOption.IGNORE_CASE)
private val MARKDOWN_IMAGE = Regex("""!\[([^\]]*)]\((https://[^\s)]+)(?:\s+["'][^"']*["'])?\)""")
private val TRAILING_SPACES = Regex("""[ \t]+\n""")
private val BLANK_RUNS = Regex("""\n{3,}""")
private val DEVIN_SESSION_ID = Regex("""^devin-[A-Za-z0-9_-]+$""")

private val KNOWN_TYPES =
  setOf("text", "metric", "chart", "tasks", "highlight", "image", "bookmark", "devin", "first_focus", "confirm", "question")

private val json = Json { ignoreUnknownKeys = true; isLenient = false }

/** Splits Bee's reply into conversational copy and validated UI components. */
fun extractBeeUi(text: String): BeeUiExtraction {
  val components = ArrayList<BeeUiComponent>()
  var spoken =
    BEEUI_FENCE.replace(text) { match ->
      components += parseBeeUiBlock(match.groupValues[1].trim())
      ""
    }
  spoken =
    MARKDOWN_IMAGE.replace(spoken) { match ->
      val alt = match.groupValues[1]
      val url = match.groupValues[2]
      if (components.none { it is BeeUiComponent.Image && it.url == url }) {
        components += BeeUiComponent.Image(url, alt.trim().ifEmpty { "Generated image" })
      }
      ""
    }
  spoken = spoken.replace(TRAILING_SPACES, "\n").replace(BLANK_RUNS, "\n\n").trim()
  return BeeUiExtraction(scrubIdentifiers(spoken), components)
}

/** Parses one fenced beeui JSON payload. Empty list when the block is malformed. */
fun parseBeeUiBlock(payload: String): List<BeeUiComponent> {
  val envelope = runCatching { json.parseToJsonElement(payload) }.getOrNull() as? JsonObject ?: return emptyList()
  val list = envelope["components"] as? JsonArray ?: return emptyList()
  val components = ArrayList<BeeUiComponent>()
  for (value in list) {
    val obj = value as? JsonObject ?: return emptyList()
    val type = obj.str("type") ?: return emptyList()
    if (type !in KNOWN_TYPES) {
      components += BeeUiComponent.Unsupported
      continue
    }
    components += parseComponent(type, obj) ?: return emptyList()
  }
  return components
}

private fun parseComponent(type: String, o: JsonObject): BeeUiComponent? =
  when (type) {
    "text" -> BeeUiComponent.Text(scrubIdentifiers(o.str("body") ?: return null))
    "metric" ->
      BeeUiComponent.Metric(
        label = scrubIdentifiers(o.str("label") ?: return null),
        value = scrubIdentifiers(o.str("value") ?: return null),
        delta = o.optStr("delta") { return null }?.let(::scrubIdentifiers),
      )
    "chart" -> {
      if (o.str("kind") != "bar") return null
      val data =
        (o["data"] as? JsonArray ?: return null).map { item ->
          val bar = item as? JsonObject ?: return null
          BeeUiComponent.Chart.Bar(bar.str("label") ?: return null, bar.num("value") ?: return null)
        }
      if (data.isEmpty()) return null
      BeeUiComponent.Chart(scrubIdentifiers(o.str("title") ?: return null), o.optStr("unit") { return null }, data)
    }
    "tasks" -> {
      val items =
        (o["items"] as? JsonArray ?: return null).map { item ->
          val task = item as? JsonObject ?: return null
          BeeUiComponent.Tasks.Item(
            id = task.str("id") ?: return null,
            title = task.str("title") ?: return null,
            done = task.bool("done") ?: return null,
            due = task.optStr("due") { return null },
          )
        }
      BeeUiComponent.Tasks(scrubIdentifiers(o.str("title") ?: return null), items)
    }
    "highlight" ->
      BeeUiComponent.Highlight(scrubIdentifiers(o.str("title") ?: return null), scrubIdentifiers(o.str("body") ?: return null))
    "image" ->
      BeeUiComponent.Image(
        url = o.httpsUrl("url") ?: return null,
        alt = scrubIdentifiers(o.str("alt") ?: return null),
        title = o.optStr("title") { return null }?.let(::scrubIdentifiers),
      )
    "bookmark" -> {
      val kind = o.optStr("kind") { return null }
      if (kind != null && kind !in setOf("website", "tweet", "youtube")) return null
      val labels =
        when (val raw = o["labels"]) {
          null, JsonNull -> null
          is JsonArray -> raw.map { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content ?: return null }.also { if (it.size > 8) return null }
          else -> return null
        }
      BeeUiComponent.Bookmark(
        title = scrubIdentifiers(o.str("title") ?: return null),
        url = o.httpsUrl("url") ?: return null,
        kind = kind,
        labels = labels,
        note = o.optStr("note") { return null }?.let(::scrubIdentifiers),
      )
    }
    "devin" -> {
      val sessionId = o.str("sessionId") ?: return null
      if (!DEVIN_SESSION_ID.matches(sessionId)) return null
      val pullRequests =
        (o["pullRequests"] as? JsonArray ?: return null).map { item ->
          val pr = item as? JsonObject ?: return null
          BeeUiComponent.Devin.PullRequest(pr.httpsUrl("url") ?: return null, pr.optStr("state") { return null })
        }
      if (pullRequests.size > 20) return null
      BeeUiComponent.Devin(
        title = scrubIdentifiers(o.str("title") ?: return null),
        status = o.str("status") ?: return null,
        statusDetail = o.optStr("statusDetail") { return null },
        sessionId = sessionId,
        sessionUrl = o.httpsUrl("sessionUrl") ?: return null,
        summary = o.optStr("summary") { return null }?.let(::scrubIdentifiers),
        pullRequests = pullRequests,
      )
    }
    "first_focus" -> {
      val expires =
        when (val raw = o["highlightExpiresAt"]) {
          null, JsonNull -> null
          else -> {
            val value = (raw as? JsonPrimitive)?.doubleOrNull ?: return null
            if (!value.isFinite() || value < -8_640_000_000_000_000.0 || value > 8_640_000_000_000_000.0) return null
            value
          }
        }
      BeeUiComponent.FirstFocus(
        requestId = o.str("requestId")?.takeIf { it.isNotEmpty() } ?: return null,
        goalTitle = o.str("goalTitle")?.takeIf { it.isNotEmpty() } ?: return null,
        projectTitle = o.str("projectTitle")?.takeIf { it.isNotEmpty() } ?: return null,
        taskTitle = o.str("taskTitle")?.takeIf { it.isNotEmpty() } ?: return null,
        seed = o.optStr("seed") { return null }?.also { if (it.isEmpty()) return null },
        highlightExpiresAt = expires,
      )
    }
    "confirm" -> {
      val payload =
        when (val raw = o["payload"]) {
          null, JsonNull -> null
          is JsonObject -> raw
          else -> return null
        }
      BeeUiComponent.Confirm(scrubIdentifiers(o.str("summary") ?: return null), o.str("action") ?: return null, payload)
    }
    "question" -> parseQuestion(o)
    else -> null
  }

private fun parseQuestion(o: JsonObject): BeeUiComponent.Question? {
  val prompts =
    (o["questions"] as? JsonArray ?: return null).map { item ->
      val prompt = item as? JsonObject ?: return null
      val header = prompt.str("header")?.trim()?.takeIf { it.length in 1..24 } ?: return null
      val question = prompt.str("question")?.trim()?.takeIf { it.length in 1..180 } ?: return null
      val options =
        when (val raw = prompt["options"]) {
          null, JsonNull -> null
          is JsonArray ->
            raw
              .map { option ->
                val obj = option as? JsonObject ?: return null
                val label = obj.str("label")?.trim()?.takeIf { it.length in 1..40 } ?: return null
                val description = obj.optStr("description") { return null }?.trim()?.also { if (it.length !in 1..120) return null }
                BeeUiComponent.Question.Option(scrubIdentifiers(label), description?.let(::scrubIdentifiers))
              }
              .also { if (it.size !in 2..3) return null }
          else -> return null
        }
      BeeUiComponent.Question.Prompt(scrubIdentifiers(header), scrubIdentifiers(question), options)
    }
  if (prompts.size !in 1..3) return null
  return BeeUiComponent.Question(prompts)
}

/**
 * Derives the actionable follow-ups from one reply. A Web3 confirmation is
 * surfaced only when exactly one action-bound confirm card is present.
 */
fun deriveBeeUiFollowUps(components: List<BeeUiComponent>): BeeUiFollowUps {
  val firstFocus = components.filterIsInstance<BeeUiComponent.FirstFocus>().firstOrNull()
  val web3 = components.mapNotNull { (it as? BeeUiComponent.Confirm)?.web3Confirmation() }
  val reversed = components.asReversed()
  val confirmation = reversed.filterIsInstance<BeeUiComponent.Confirm>().firstOrNull { it.web3Confirmation() == null }
  val question = reversed.filterIsInstance<BeeUiComponent.Question>().firstOrNull()
  return BeeUiFollowUps(
    firstFocus = firstFocus,
    web3Confirmation = web3.singleOrNull(),
    confirmationSummary = confirmation?.summary,
    question = question,
  )
}

/** A confirm card is action-bound only through a non-empty string `payload.web3ActionId`. */
fun BeeUiComponent.Confirm.web3Confirmation(): Web3Confirmation? {
  val actionId = (payload?.get("web3ActionId") as? JsonPrimitive)?.takeIf { it.isString }?.content?.takeIf { it.isNotEmpty() }
  return actionId?.let { Web3Confirmation(it, summary) }
}

/** Echoes a question-card choice back to Bee as a normal chat message. */
fun questionAnswer(question: String, answer: String): String = "For \u201C$question\u201D, my answer is \u201C$answer\u201D."

/** Hostname without the `www.` prefix; falls back to the raw URL text. */
fun bookmarkHost(url: String): String =
  runCatching { URI(url).host?.removePrefix("www.") }.getOrNull() ?: url

/** Suggested file name when saving a generated image. */
fun generatedImageFileName(url: String, now: Long = System.currentTimeMillis()): String {
  val sourceName = runCatching { URI(url).path.substringAfterLast('/') }.getOrNull() ?: ""
  return if (Regex("""\.(?:avif|gif|jpe?g|png|webp)$""", RegexOption.IGNORE_CASE).containsMatchIn(sourceName)) sourceName
  else "bee-image-$now.png"
}

private fun JsonObject.str(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

/** Optional string: absent or null passes as null; any other non-string type fails. */
private inline fun JsonObject.optStr(key: String, onInvalid: () -> Nothing): String? =
  when (val raw = this[key]) {
    null, JsonNull -> null
    is JsonPrimitive -> if (raw.isString) raw.content else onInvalid()
    else -> onInvalid()
  }

private fun JsonObject.num(key: String): Double? = (this[key] as? JsonPrimitive)?.takeIf { !it.isString }?.doubleOrNull?.takeIf { it.isFinite() }

private fun JsonObject.bool(key: String): Boolean? = (this[key] as? JsonPrimitive)?.takeIf { !it.isString }?.booleanOrNull

private fun JsonObject.httpsUrl(key: String): String? {
  val value = str(key) ?: return null
  if (!value.startsWith("https://")) return null
  val uri = runCatching { URI(value) }.getOrNull() ?: return null
  if (uri.host.isNullOrEmpty()) return null
  return value
}
